File: cpp/Stack/Expression.cpp
#include <iostream>
#include "Expression.h"
#include <stack>
#include <string>

using namespace std;


char* stringToCharArray(std::string str) {
    int n = str.length();
    char char_array[n + 1];
    strcpy(char_array, str.c_str());
}


int indexOfChar(char* arr, char ch) {
    size_t len = strlen(arr);
    for (int i = 0; i < len; ++i) {
        if (arr[i] == ch)
            return i;
    }
    return -1;
}


bool Expression::isLeftBracket(char ch) {
    for (int i = 0; i < kElementsOfBracket; i++) {
        if (ch == leftBrackets[i])
            return true;
    }
    return false;
}


bool Expression::isRightBracket(char ch) {
    for (int i = 0; i < kElementsOfBracket; i++) {
        if (ch == rightBrackets[i])
            return true;
    }
    return false;
}


bool Expression::isBracketsMatched(char left, char right) {

    return indexOfChar(leftBrackets, left) == indexOfChar(rightBrackets, right);
}


bool Expression::isBalanced(std::string input) {
    std::stack<char> result;

    for (int i = 0; i < input.size(); i++) {
        char ch = input[i];

        if (isLeftBracket(ch)) result.push(ch);

        else if (isRightBracket(ch)) {

            if (result.empty()) return false;
        
            char top = result.top();
            result.pop();
            if (!isBracketsMatched(top, ch)) return false;
        }
    }
    return result.empty();

}

File: python/Stack/Expression.py
class Expression:
    def __init__(self):
        #super().__init__()
        self.k_elements_of_bracket = 4
        self.__left_brackets = ['(', '{', '<', '[']
        self.__right_brackets = [')', '}', '>', ']']
    
    def __is_left_bracket(self, ch):
        return ch in self.__left_brackets
    
    def __is_right_bracket(self, ch):
        return ch in self.__right_brackets
    
    def __is_bracket_matched(self, left, right):
        return self.__left_brackets.index(left) == self.__right_brackets.index(right)
    
    def _is_balanced(self, string):
        stack = []
        for ch in string:
            if self.__is_left_bracket(ch):
                stack.append(ch)
                
            elif self.__is_right_bracket(ch):
                if len(stack) == 0:
                    return False
                
                top = stack.pop(-1)
                if not self.__is_bracket_matched(top, ch):
                    return False
                
        return len(stack) == 0
                
                
if __name__ == '__main__':
    exp = Expression()

    string = "()]"
    print(exp._is_balanced(string))
    